<a href="https://cognitiveclass.ai"><img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width = 400> </a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>


## Capstone Project for Introduction to Deep Learning with Keras


In this project, you will build a regression model using the Keras library to model the same data about concrete compressive strength that we used in labs 3. For your convenience, the data can be found here again: https://cocl.us/concrete_data. To recap, the predictors in the data of concrete strength include:

Cement
Blast Furnace Slag
Fly Ash
Water
Superplasticizer
Coarse Aggregate
Fine Aggregate



The four parts of the capstone:
<h5> A. Build a baseline model</h5>
<h5> B. Normalize the data </h5>
<h5> C. Increate the number of epochs </h5>  
<h5> D. Increase the number of hidden layers </h5>  



<h2>Regression Models with Keras</h2>

<h3>Capstone Project<h3>    
<h5> 1. Download and Clean dataset</h5>
<h5> 2. Build a Regression Model with Keras </h5>
<h5> 3. Train and Test the Network. </h5>    


## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#item31">Download and Clean Dataset</a>  
2. <a href="#item32">Import Keras</a>  
3. <a href="#item33">Build a Neural Network</a>  
4. <a href="#item34">Train and Test the Network</a>  

</font>
</div>


<a id="item31"></a>


## Download and Clean Dataset


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [17]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [18]:
# Download Data
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 


#### Let's check how many data points we have.


In [19]:
# Checking size of data
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [20]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [21]:
# Checking for Null values
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [22]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # input data or predictors drawn form data
target = concrete_data['Strength'] # set target to strenght

<a id="item2"></a>


Let's do a quick sanity check of the predictors and the target dataframes.


In [23]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [24]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.


In [50]:
#

Let's save the number of predictors to _n_cols_ since we will need this number when building our network.


In [28]:
n_cols = predictors_norm.shape[1] # number of predictors
print(n_cols)

8


<a id="item1"></a>


<a id='item32'></a>


## Import Keras


Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.


#### Let's go ahead and import the Keras library


In [29]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.


Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.


In [30]:
from keras.models import Sequential
from keras.layers import Dense

## 1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_splithelper function from Scikit-learn


In [34]:
from sklearn.model_selection import train_test_split
predictors_train, predictors_test, target_train, target_test = train_test_split( predictors, target, test_size=0.3, random_state=4)
print ('Train set:', predictors_train.shape,  target_train.shape)
print ('Test set:', predictors_test.shape,  target_test.shape)

Train set: (721, 8) (721,)
Test set: (309, 8) (309,)


<a id='item33'></a>


## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

One hidden layer of 10 nodes, and a ReLU activation function

In [37]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

In [47]:
%whos

Variable                Type          Data/Info
-----------------------------------------------
Dense                   type          <class 'keras.layers.core.Dense'>
Sequential              type          <class 'keras.models.Sequential'>
concrete_data           DataFrame           Cement  Blast Furna<...>\n[1030 rows x 9 columns]
concrete_data_columns   Index         Index(['Cement', 'Blast F<...>],\n      dtype='object')
keras                   module        <module 'keras' from '/ho<...>kages/keras/__init__.py'>
l                       list          n=27
mean_squared_error      function      <function mean_squared_error at 0x7fd0ac728b70>
model                   Sequential    <keras.models.Sequential <...>object at 0x7fd00c6778d0>
mse                     float64       119.9451045159662
n_cols                  int           8
np                      module        <module 'numpy' from '/ho<...>kages/numpy/__init__.py'>
pd                      module        <module 'pandas' from '/h<.

<a id="item4"></a>


<a id='item34'></a>


## 2. Train and Test the Network


Let's call the function now to create our model.


In [42]:
# build the model
model = regression_model()

Train the model on the training data using 50 epochs.

In [43]:
# fit the model
model.fit(predictors, target, validation_split=0.3, epochs=50, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 0s - loss: 19679.5107 - val_loss: 12132.9323
Epoch 2/50
 - 0s - loss: 7073.5429 - val_loss: 4020.2635
Epoch 3/50
 - 0s - loss: 2154.9625 - val_loss: 1661.2707
Epoch 4/50
 - 0s - loss: 1133.7264 - val_loss: 1308.6251
Epoch 5/50
 - 0s - loss: 1014.8811 - val_loss: 1232.1943
Epoch 6/50
 - 0s - loss: 967.0948 - val_loss: 1173.4271
Epoch 7/50
 - 0s - loss: 918.0373 - val_loss: 1120.7351
Epoch 8/50
 - 0s - loss: 871.6712 - val_loss: 1060.4884
Epoch 9/50
 - 0s - loss: 827.8394 - val_loss: 1007.1295
Epoch 10/50
 - 0s - loss: 786.1199 - val_loss: 956.3112
Epoch 11/50
 - 0s - loss: 746.8405 - val_loss: 899.2072
Epoch 12/50
 - 0s - loss: 709.6054 - val_loss: 852.7871
Epoch 13/50
 - 0s - loss: 675.1687 - val_loss: 809.5686
Epoch 14/50
 - 0s - loss: 643.9426 - val_loss: 765.9246
Epoch 15/50
 - 0s - loss: 613.3559 - val_loss: 722.1425
Epoch 16/50
 - 0s - loss: 586.8346 - val_loss: 684.9157
Epoch 17/50
 - 0s - loss: 559.7753 - val_loss: 645.

In [44]:
# Evaluate the model on the test data using `evaluate`
print("Evaluate on test data")
results = model.evaluate(predictors_test, target_test, batch_size=128)
print("test loss, test acc:", results)

# Generate predictions (probabilities -- the output of the last layer)
# on new data using `predict`
print("Generate predictions for 3 samples")
predictions = model.predict(predictors_test[:3])
print("predictions shape:", predictions.shape)

Evaluate on test data
309/309 [==============================] - 0s 12us/step
test loss, test acc: 217.61868602172456
Generate predictions for 3 samples
predictions shape: (3, 1)


## 4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

In [48]:
from sklearn.metrics import mean_squared_error
times = 50
l = []

for time in range(times):
    print("Number of time: ", time+1)
    predictors_train, predictors_test, target_train, target_test = train_test_split( predictors, target, test_size=0.3, random_state=4)
    model.fit(predictors, target, validation_split=0.3, epochs=50, verbose=2)
    
    pred = model.predict(predictors_test)

    mse = mean_squared_error(target_test, pred)
    l.append(mse)
    print(" ")

Number of time:  1
Train on 721 samples, validate on 309 samples
Epoch 1/50
 - 0s - loss: 126.8802 - val_loss: 73.7285
Epoch 2/50
 - 0s - loss: 125.5552 - val_loss: 73.3289
Epoch 3/50
 - 0s - loss: 125.6908 - val_loss: 73.2771
Epoch 4/50
 - 0s - loss: 127.7068 - val_loss: 76.5371
Epoch 5/50
 - 0s - loss: 130.5771 - val_loss: 73.5826
Epoch 6/50
 - 0s - loss: 125.0434 - val_loss: 72.9846
Epoch 7/50
 - 0s - loss: 124.1730 - val_loss: 77.7439
Epoch 8/50
 - 0s - loss: 129.4372 - val_loss: 76.7375
Epoch 9/50
 - 0s - loss: 125.7344 - val_loss: 73.3907
Epoch 10/50
 - 0s - loss: 125.5633 - val_loss: 74.5616
Epoch 11/50
 - 0s - loss: 126.3566 - val_loss: 73.1163
Epoch 12/50
 - 0s - loss: 126.0428 - val_loss: 73.5532
Epoch 13/50
 - 0s - loss: 127.1601 - val_loss: 73.9827
Epoch 14/50
 - 0s - loss: 128.1026 - val_loss: 73.0475
Epoch 15/50
 - 0s - loss: 127.6863 - val_loss: 72.7587
Epoch 16/50
 - 0s - loss: 129.4636 - val_loss: 95.7959
Epoch 17/50
 - 0s - loss: 132.4276 - val_loss: 79.6324
Epoch 18/

## 5. Report the mean and the standard deviation of the mean squared errors.

In [49]:
mean = np.mean(l)
std = np.std(l)

print("Mean: ", mean)
print("Standard Deviation: ", std)

Mean:  95.00060315109855
Standard Deviation:  31.552997563895342


This notebook is part of a course on **Coursera** called _Introduction to Deep Learning & Neural Networks with Keras_
